tee-worker/litentry/core/vc-task/receiver/src/lib.rs

```diff
@@ -154,30 +171,145 @@ where
 	N: AccessNodeMetadata + Send + Sync + 'static,
 	N::MetadataType: NodeMetadataTrait,
 {
-	let start_time = Instant::now();
-	let enclave_shielding_key = context
-		.shielding_key
-		.retrieve_key()
-		.map_err(|e| format!("Failed to retrieve shielding key: {:?}", e))?;
-	let tcs = request
+	let request = &mut vc_req.request;
+	let sender = &vc_req.sender;
+	let enclave_shielding_key = match context.shielding_key.retrieve_key() {
+		Ok(value) => value,
+		Err(e) => {
+			send_vc_response(
+				sender,
+				context,
+				Err(format!("Failed to retrieve shielding key: {:?}", e)),
+				0u8,
+				1u8,
+			);
+			return
+		},
+	};
+	let tcs = match request
 		.decrypt(Box::new(enclave_shielding_key))
 		.ok()
 		.and_then(|v| TrustedOperation::<TrustedCallSigned, Getter>::decode(&mut v.as_slice()).ok())
 		.and_then(|top| top.to_call().cloned())
-		.ok_or_else(|| "Failed to decode payload".to_string())?;
-
+	{
+		Some(tcs) => tcs,
+		None => {
+			send_vc_response(
+				sender,
+				context,
+				Err("Failed to decode request payload".to_string()),
+				0u8,
+				1u8,
+			);
+			return
+		},
+	};
 	let mrenclave = match context.ocall_api.get_mrenclave_of_self() {
 		Ok(m) => m.m,
-		Err(_) => return Err("Failed to get mrenclave".to_string()),
+		Err(_) => {
+			send_vc_response(sender, context, Err("Failed to get mrenclave".to_string()), 0u8, 1u8);
+			return
+		},
 	};
+	if !tcs.verify_signature(&mrenclave, &request.shard) {
+		send_vc_response(sender, context, Err("Failed to verify sig".to_string()), 0u8, 1u8);
+		return
+	}
+	if let TrustedCall::request_vc(..) = tcs.call {
+		let response = process_single_request(
+			request.shard,
+			context.clone(),
+			extrinsic_factory,
+			node_metadata_repo,
+			tc_sender,
+			tcs.call.clone(),
+		);
+		send_vc_response(sender, context, response, 0u8, 1u8);
+	} else if let TrustedCall::request_batch_vc(signer, who, assertions, maybe_key, req_ext_hash) =
+		tcs.call
+	{
+		// Filter out duplicate assertions
+		let mut seen: HashSet<H256> = HashSet::new();
+		let mut unique_assertions = Vec::new();
+		for assertion in assertions.into_iter() {
+			let hash = H256::from(blake2_256(&assertion.encode()));
+			if seen.insert(hash) {
+				unique_assertions.push(assertion);
+			}
+		}
```

Comment:
I think HashSet itself is enough to dedup, something like `let set: HashSet<_> = vec.drain(..).collect();`

Reply:
OK not if you want to find out the duplicated assertion and return the error for it...